### PR TITLE
Initial Security Model

### PR DESCRIPTION
**Pull request by pjbgf**
#### What type of PR is this?

/kind documentation

#### What this PR does / why we need it:

Initial security model markdown to inform users on permissions, and overall security related information.
In the future, this will contain also include compliance status (i.e. NIST, SLSA, etc.)

#### Which issue(s) this PR fixes:

Relates to #653

#### Does this PR have test?

N/A

#### Special notes for your reviewer:

N/A

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

**Comment by pjbgf** [comment]
@jhrozek @JAORMX @saschagrunert keen on hearing thoughts on how we can improve this. Also, need some other pairs of eyes to double check on the selinux front.

**Review by saschagrunert — APPROVED** [review]
I love it, it’s great to have these docs!

Do we plan in the future to only distinguish those two modes by configuration or do we wanna keep the operator configurable as is?

**Comment by saschagrunert** [comment]
/hold
For other reviews.

**Review by jhrozek — COMMENTED** [review]
This is really great work. I left some comments inline.

Let's get this merged soon and then I'll add some OCP specific info on top of that (e.g. we are using the privileged SCC ...)

Btw does it make sense to mention that by default almost all the containers run with the `spc_t` SELinux type which pretty much means "unconfined"? (If you prefer, I can propose another PR atop this)

**Comment by pjbgf** [comment]
> Btw does it make sense to mention that by default almost all the containers run with the `spc_t` SELinux type which pretty much means "unconfined"? (If you prefer, I can propose another PR atop this)

@jhrozek great point - yes. I think we should get out at a container level:
- SELinux type
- AppArmor profile
- Seccomp profile
- Capabilities

Happy for you to propose a follow-up PR on that. :+1: 

On the mid/long term, it would be great if we could think on ways to automatically extract such information (like we do for RBAC).

File: security-model.md
Comment (pjbgf) [reply security-model.md:36]: The idea here was to be more generic, meaning that for SELinux feature to be operational, something must be running at that given permission level to be able support it.

Initially I thought as marking with `No*` and adding a caveat soon after, but then decided to make it easier. But I agree, it is not completely clear. 
Do you think it would be better if we broke down each of the components that support each enforcement technology (selinux, apparmor, selinux)?

File: security-model.md
Comment (pjbgf) [reply security-model.md:37]: Sure, let me fix that - SELinux. 

For AppArmor, it does currently as it needs to enter the host `mnt` namespace to gain access to the AppArmorFS on demand. There are other ways of doing it, but that's the current approach.

File: security-model.md
Comment (pjbgf) [reply security-model.md:59]: Fair, to be honest I think we need to add a lot more meat to this section. But was thinking on leaving as is and improve on follow-up PRs. 

Overall, profile generation is tricky and generally can be done either with target workload being a) unrestricted or b) restricted with existing/template profile. The former may have security implications and the latter would yield a partial profile - assuming that those execution paths may not be fully materialised when restricted.

In my point of view, we can have here a disclaimer of potential risks, but then have a full on section on profiling in which we suggest how to do that correctly and securely. Would you be OK with that?

File: security-model.md
Comment (pjbgf) [reply security-model.md:66]: Fair point, this was just a reference to help end users to compare SPO with other projects that provide similar capabilities.

Some of those projects require SSH credentials in order to executec commands in the nodes to install/remove profiles.

File: security-model.md
Comment (pjbgf) [reply security-model.md:39]: Same as https://github.com/kubernetes-sigs/security-profiles-operator/pull/693#discussion_r756985410

In the future we could potentially have a page to compare with other projects and help folks decide. But that may be hard to keep up to date.

File: security-model.md
Comment (pjbgf) [reply security-model.md]: Fixed it accordingly.

**Comment by pjbgf** [comment]
@jhrozek went through your poinds - thank you so much for the extensive feedback. PTAL

File: security-model.md
Comment (pjbgf) [reply security-model.md]: Sure, going through it again it does not come across as clear as I intented it to be.

On a follow-up PR I think this can be replaced by breaking down the "host paths" for each main feature (SAS) and the access (rw) they require. So regardless of this is implemented, for Seccomp to work it requires `rw` access to `{kubelet_path}/seccomp`. And the same for selinux/apparmor.

File: security-model.md
Comment (saschagrunert) [inline security-model.md:52]: There are a bunch of more host paths, should we cover them as well: https://cs.k8s.io/?q=HostPathVolumeSource&i=nope&files=&excludeFiles=vendor&repos=kubernetes-sigs/security-profiles-operator

**Comment by saschagrunert** [comment]
> @saschagrunert thank you. In my opinion we can keep as is right now. Once we are a bit more settled in terms of features - specially around AppArmor and the great bits around eBPF profiling you are working on - we review it. What do you think?

Sounds great!

File: security-model.md
Comment (pjbgf) [reply security-model.md:52]: @saschagrunert nice! I did not know about `https://cs.k8s.io` :+1: 

Yes, I think we should break it all down and make it explicit what we need and to what level.
This will also help when we create our AppArmor profiles - which could actually become our official documentation on host-level disk permissions.

File: security-model.md
Comment (jhrozek) [reply security-model.md]: I think your approach is better and we might want to just add a side-note or a sub-paragraph later.

File: security-model.md
Comment (jhrozek) [reply security-model.md]: Ah, I know what confused me, the code that uses hostPIDs for AppArmor is not merged yet, but it is part of PR#680 right?

File: security-model.md
Comment (jhrozek) [reply security-model.md]: Right, selinuxd needs to write to the hosts's selinux db. Thanks for the explanation.

File: security-model.md
Comment (jhrozek) [reply security-model.md]: Oh I didn't know that. Leave it in then, I like to show off where SPO is better!

**Comment by jhrozek** [comment]
I'm happy with the current state!
/lgtm

File: security-model.md
Comment (pjbgf) [reply security-model.md]: yes, that's correct. once that is merged and apparmor is enabled, spod will require HostPID.

**Review by saschagrunert — APPROVED** [review]

File: security-model.md
Comment (JAORMX) [inline security-model.md:43]: one thing to note (maybe this can go in a separate PR) is that when SELinux is enabled, the SPOd has a sidecar container running selinuxd. selinuxd actually runs with a `selinux.process` type and is not a privileged container. Although it does run as root. Something we could start investigating is trying to run it with a non-root user and seeing if we can make do with enabling extra capabilities.... However, a lot of the attack surface of that container  is mitigated by the custom policy we use for it.

File: security-model.md
Comment (JAORMX) [inline security-model.md:51]: we've removed this now https://github.com/kubernetes-sigs/security-profiles-operator/pull/698

**Review by JAORMX — APPROVED** [review]
Looks like a good start! Thanks for setting this up!

**Pull request merged**
